src/attr/_make.py

```diff
@@ -2062,7 +2081,27 @@ def _make_repr(attrs, ns, cls):
         "    else:",
         "      already_repring.add(id(self))",
         "  try:",
-        f"    return f'{cls_name_fragment}({repr_fragment})'",
+        f"     if not {only_non_default_attr_in_repr}:",
+        f"         return f'{cls_name_fragment}({repr_fragment})'",
+        "     attr_frags = []",
+        "     for a in getattr(self, '__attrs_attrs__', []):",
+        "         value = getattr(self, a.name, NOTHING)",
+        "         if hasattr(a.default, 'factory') and callable(a.default.factory):",
+        "             if a.default.takes_self:",
+        "                 default_ = a.default.factory(self)",
+        "             else:",
+        "                 default_ = a.default.factory()",
+        "         else:",
+        "             default_ = a.default",
```

Comment:
* I think you could generate the defaults only once on first repr and cache them for subsequent runs.
* IIRC, attrs applies converters on defaults, so shouldn't this do, too?
* Calling factories unconditionally is a bit problematic for things like `factory=lambda: next(generator)` because it would advance the caller's generator without an actual instantiation. Overall, it might be better to keep track of the defaults produced on construction and compare against that?

FWIW, I had written a [caller-side wrapper to hide defaults](https://gist.github.com/mara004/5f02da5f3a09f9206c0755b5bcf6f105). (The way it works around the factory problem is by letting the caller mark factories as dynamic if they don't want them called in repr.)

Disclaimer: I'm not an attrs maintainer, so my views may not matter here.